pallets/pallet-bonded-coins/src/try_state.rs

```diff
@@ -0,0 +1,83 @@
+use frame_support::traits::{
+	fungible::InspectHold,
+	fungibles::{metadata::Inspect as InspectMetadata, roles::Inspect as InspectRoles, Inspect},
+};
+use sp_runtime::{traits::Zero, TryRuntimeError};
+
+use crate::{types::PoolDetails, Config, FungiblesAssetIdOf, HoldReason, Pools};
+
+pub(crate) fn do_try_state<T: Config>() -> Result<(), TryRuntimeError> {
+	// checked currency ids. Each Currency should only be associated with one pool.
+	let mut checked_currency_ids = Vec::<FungiblesAssetIdOf<T>>::new();
+
+	Pools::<T>::iter().try_for_each(|(pool_id, pool_details)| {
+		let PoolDetails {
+			collateral_id,
+			deposit,
+			owner,
+			bonded_currencies,
+			state,
+			denomination,
+			..
+		} = pool_details;
+
+		let pool_account = pool_id.into();
+
+		// Collateral checks
+		let collateral_exists = T::Collaterals::asset_exists(collateral_id.clone());
+		assert!(collateral_exists);
+		let collateral_issuance = T::Collaterals::total_issuance(collateral_id);
+
+		// Deposit checks
+		let balance_on_hold_user =
+			T::DepositCurrency::balance_on_hold(&T::RuntimeHoldReason::from(HoldReason::Deposit), &owner);
+		assert!(balance_on_hold_user >= deposit);
+
+		// Bonded currencies checks
+		bonded_currencies
+			.iter()
+			.try_for_each(|currency_id| -> Result<(), TryRuntimeError> {
+				// check if currency is already associated with another pool
+				if checked_currency_ids.contains(currency_id) {
+					return Err(TryRuntimeError::Other(
+						"Currency is already associated with another pool",
+					));
+				}
+
+				checked_currency_ids.push(currency_id.clone());
```

Comment:
If you use a hashset as Raphael mentions above, you can simply try to insert the new currency ID, and check for the `bool` result which tells you whether an asset with the same ID was already in the set, which would result in an error. Less lines of code and easier to understand. Also the use of a hash set makes it clear that we expect unique elements in there.

Reply:
The asset ID must implement Ord, which it currently doesn’t. I believe the changes required to make this work are not worth the effort.